cmds-42/usr.bin/rcs/man/ident.1
.TH IDENT L "October 30, 1989" "Purdue University"
.SH NAME
ident \- identify files
.SH SYNOPSIS
\fBident\fR [ \fB\-q\fR ] [ file ... ]
.SH DESCRIPTION
.I Ident
searches the named files or, if no file name appears, the standard input
for all occurrences of the pattern
\fI$keyword:...$\fR, where \fIkeyword\fR is one of
.nf

        Author
        Date
        Header
        Id
        Locker
        Log
        Revision
        RCSfile
        Source
        State

.fi
These patterns are normally inserted automatically by the RCS command
.IR co (L),
but can also be inserted manually. The option \fB\-q\fR suppresses
the warning given if there are no patterns in a file.
.PP
\fIIdent\fR works on text files as well as object files and dumps.
For example, if the C program in file f.c contains
.nf

        char rcsid[] = "$\&Header:  Header information $";

.fi
and f.c is compiled into f.o, then the command
.nf

        ident  f.c  f.o

will print

        f.c:
                $\&Header:  Header information $
        f.o:
                $\&Header:  Header information $

.SH IDENTIFICATION
.de VL
\\$2
..
Author: Walter F. Tichy,
Purdue University, West Lafayette, IN, 47907.
.sp 0
Revision Number:
.VL $Revision: 1.3 $
; Release Date:
.VL $Date: 89/10/30 17:36:36 $
\&.
.sp 0
Copyright \(co 1982, 1988, 1989 by Walter F. Tichy.
.SH SEE ALSO
ci(L), co(L), rcs(L), rcsdiff(L), rcsintro(L), rcsmerge(L), rlog(L), rcsfile(L),
.sp 0
Walter F. Tichy, "Design, Implementation, and Evaluation of a Revision Control
System," in \fIProceedings of the 6th International Conference on Software
Engineering\fR, IEEE, Tokyo, Sept. 1982.


cmds-42/ChangeLog
usr.ucb/from/from.c				mmeyer 8/22/89

Change from to do the from on geteuid(), instead of getlogin() which
looks at who is in /etc/utmp.  

usr.ucb/telnet/telnet.c				mmeyer 8/22/89

Fix for bug #2889.

bin/sed/sed1.c					mmeyer 8/21/89

Fix for bug #2831.

usr.bin/calendar/calendar.{c,sh}		mmeyer 8/21/89

Fix for bug #2794

usr.ucb/Mail					mmeyer 6/27/89

Upgraded to Tahoe sources and fixed a few bugs.

etc/mount/mount.c				mmeyer 6/26/89

Change reversal of mounting of nfs partitions.

etc/init/init.c					mmeyer 6/23/89

Changes for onerror/onoption.

etc/rlogind/rlogind.c				mmeyer 6/14/89

Took out malloc from signal handler.

etc/nu/*					mmeyer 6/13/89

Lee's new nu.

etc/ifconfig/ifconfig.c				mmeyer 6/13/89
etc/bootpd/*

Changes for sexy netinit

bin/awk/awk.g.y					mmeyer 6/1/89

Fixed core dump for Trey.  

etc/quot/quot.c					mmeyer 5/31/89

Changed acct() to print out the correct number of 1024 byte blocks.

bin/sh/homedir.c				mmeyer 5/26/89

Changed sh to use getpwnam() instead of searching through /etc/passwd

usr.ucb/indent/lexi.c				wrp    5/22/89
usr.ucb/indent/indent.c		
usr.ucb/indent/indent_globs.h

Bill Parkhurst's changes for Objective-C 4.0 support, along with
a bugfix that Brian and Morris made.


bin/diff/diffh_vers.c				mike   5/22/89
usr.bin/diff3/Makefile				

Bugfix from Mike DeMoney.  Fixes problem that occurs during integration.

usr.bin/tip/acu.c				mmeyer 5/20/89
usr.bin/tip/cu.c
usr.bin/tip/tip.c
usr.bin/tip/cmds.c

Changed references to connect(), abort() and send() to compile
with shlibs.

bin/csh/sh.c					mmeyer 5/19/89

Integrated the atexit() mod.

usr.ucb/rlogin/rlogin.c				mmeyer 5/19/89

Verified that Mike DeMoney's change to make chars with 0x7f was in.

bin/make/rules.c				mmeyer 5/19/89
bin/make/doname.c	

Integrated Mike's fixes.

usr.etc/mach_swapon/swaptab			mmeyer 5/19/89

Changed to /private/vm/swaptab instead of /etc/vm/swaptab

usr.bin/calendar/Makefile			mmeyer 5/18/89

Changed Makefile so /usr/lib/calendar would install correctly.

etc/route/route.c				mmeyer 5/16/89
etc/rwhod/rwhod.c
usr.etc/nfsstat/nfsstat.c
usr.etc/rpc.rstatd/rpc.rstatd.c
usr.ucb/sysline/sysline.c (not that this will ever run on our system)
usr.ucb/sendmail/src/conf.c
usr.bin/uucp/acucntrl.c

Changed all to nlist "/mach" instead of "/vmunix"

etc/syslog/syslog.c				mmeyer 5/15/89

Changed syslog to print out "mach:" instead of "vmunix:"

bin/date/date.c					mmeyer 5/14/89

Initialized tz.{tz_minuteswest,tz_dsttime} to zero.  This bug
cropped up when linking with shared libraries.  Kevin Enderby
has a bug report.

mk/cmds/usr/ucb/w/Makefile			mmeyer 5/14/89

Linked uptime to w in the INSTALL phase

usr.bin/uucp/fio.c				mmeyer 5/9/89
usr.bin/uucp/gio.c
usr.bin/uucp/logent.c
usr.bin/uucp/tio.c

Ifdef'd syslog() so uucp would link with the libsys_s shlib.
	

usr.bin/uucp/Makefile

Took out duplicate index.c from uucp.

usr.bin/troff/Makefile				mmeyer 5/9/89

Took out malloc.c and calloc.c from CFILES.  troff had its own 
version of calloc,realloc,malloc and free.

usr.bin/nroff/Makefile				mmeyer 5/9/89

Took out malloc.c and calloc.c from CFILES.  nroff had its own 
version of calloc,realloc,malloc and free.

usr.ucb/rusers/Makefile				mmeyer 5/9/89

Took out malloc.c and calloc.c from CFILES.  rusers had its own 
version of calloc,realloc,malloc and free.

usr.ucb/msgs/msgs.c	  			mmeyer 5/9/89

Ifdef'd send() so /usr/ucb/msgs would link with the libsys_s shlib.

usr.ucb/more/more.c	  			mmeyer 5/9/89

Ifdef'd strlen() so /usr/ucb/more would link with the libsys_s shlib.

usr.ucb/ex/exrecover.c  			mmeyer 5/9/89

Ifdef'd fprintf() so /usr/ucb/exrecover would link with the libsys_s shlib.

usr.ucb/Mail/aux.c  				mmeyer 5/9/89

Ifdef'd index() so /usr/ucb/Mail would link with the libsys_s shlib.

usr.ucb/Mail/send.c				mmeyer 5/9/89
usr.ucb/Mail/cmd1.c
usr.ucb/Mail/cmd2.c
usr.ucb/Mail/collect.c
usr.ucb/Mail/edit.c
usr.ucb/Mail/fio.c
usr.ucb/Mail/quit.c

Ifdef'd send() so /usr/ucb/Mail would link with the libsys_s shlib.

usr.ucb/Mail/head.c
usr.ucb/Mail/list.c
usr.ucb/Mail/aux.c

Ifdef'd raise() so /usr/ucb/Mail would link with the libsys_s shlib.

etc/sa/sa.c					mmeyer 5/9/89

Ifdef'd bcmp() so /etc/sa would link with the libsys_s shlib.

etc/routed/defs.h				mmeyer 5/9/89
etc/routed/startup.c
etc/routed/inet.c 

Ifdef'd inet_makeaddr() so /etc/routed would link with the libsys_s shlib.

bin/su/su.c					mmeyer 5/9/89

Unifdef'd getenv() so su would link with the libsys_s shlib.

bin/awk/Makefile				mmeyer 5/9/89

Took out malloc.c and calloc.c from CFILES.  awk had its own 
version of calloc,realloc,malloc and free.

All CMDS Makefiles				mmeyer 5/8/89

Changed all Makefiles to have -O -g for CFLAGS and -lsys_s for
LIBS.  Changes above are for commands that didn't link.

cmds/bin/cp/cp.c				mmeyer 5/8/89
cmds/bin/cat/cat.c

Integrated Avie's mapped file changes.  He claims a 10% performace
win with mapped files.

cmds/bin/make/main.c				mmeyer 5/1/89

Changed main() to have make to a setrlimit(RLIMIT_STACK) to trap
infinitely recursive macros instead of core dumping.

cmds/bin/sh/homedir.c				mmeyer  4/30/89

Changed sh to use getpwent() rather than grep'ing through /etc/passwd.

cmds/bin/tar/tar.c				mmeyer  4/30/89

Changed the "v" flag printf()'s to fprintf(stderr)'s

cmds/etc/icheck/icheck.c			mmeyer  4/30/89

Modified icheck to know about fast symlinks.

cmds/usr.ucb/rlogin/rlogin.c			mmeyer

Added the oob() signal handler for the SIGURG signal.  Data
was getting lost between the socket and the rlogin program.

cmds/bin/csh/sh.c				mmeyer

Added lines 1022 to 1029 in dosource() which check for no operand 
for source flag.

Changed lines 1036 to 1042 in dosource() to just print out an error
using perror() when an open of a source'd file fails instead of not
letting the login session continue.

cmds/bin/passwd/passwd.c			mmeyer

Took out a check against /bin/sh so a user can switch back
and forth from csh to sh and back again.


cmds-42/usr.bin/rcs/src/ident.c
/* Copyright (C) 1982, 1988, 1989 Walter Tichy
   Distributed under license by the Free Software Foundation, Inc.

This file is part of RCS.

RCS is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 1, or (at your option)
any later version.

RCS is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with RCS; see the file COPYING.  If not, write to
the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.

Report problems and direct all questions to:

    rcs-bugs@cs.purdue.edu

*/

/*
 *                     RCS identification operation
 */
#ifndef lint
static char rcsid[]=
"$Header: /usr/src/local/bin/rcs/src/RCS/ident.c,v 4.5 89/05/01 15:11:54 narten Exp $Purdue CS";
#endif

/* $Log:	ident.c,v $
 * Revision 4.5  89/05/01  15:11:54  narten
 * changed copyright header to reflect current distribution rules
 * 
 * Revision 4.4  87/10/23  17:09:57  narten
 * added exit(0) so exit return code would be non random
 * 
 * Revision 4.3  87/10/18  10:23:55  narten
 * Updating version numbers. Changes relative to 1.1 are actually relative
 * to 4.1
 * 
 * Revision 1.3  87/07/09  09:20:52  trinkle
 * Added check to make sure there is at least one arg before comparing argv[1]
 * with "-q".  This necessary on machines that don't allow dereferncing null
 * pointers (i.e. Suns).
 * 
 * Revision 1.2  87/03/27  14:21:47  jenkins
 * Port to suns
 * 
 * Revision 1.1  84/01/23  14:50:03  kcs
 * Initial revision
 * 
 * Revision 4.1  83/05/10  16:31:02  wft
 * Added option -q and input from reading stdin.
 * Marker matching is now done with trymatch() (independent of keywords).
 * 
 * Revision 3.4  83/02/18  17:37:49  wft
 * removed printing of new line after last file.
 *
 * Revision 3.3  82/12/04  12:48:55  wft
 * Added LOCKER.
 *
 * Revision 3.2  82/11/28  18:24:17  wft
 * removed Suffix; added ungetc to avoid skipping over trailing KDELIM.
 *
 * Revision 3.1  82/10/13  15:58:51  wft
 * fixed type of variables receiving from getc() (char-->int).
*/

#include  "rcsbase.h"
#define fflsbuf _flsbuf
/* redefinition of _flsbuf in putc not needed */
#ifndef lint
static char rcsbaseid[] = RCSBASE;
#endif

extern enum markers trymatch();

int quietflag;

main(argc, argv)
int  argc; char  *argv[];
/*  Ident searches the named files for all occurrences
 *  of the pattern $keyword:...$, where the keywords are
 *  Author, Date, Header, Id, Log, RCSfile, Revision, Source, and State.
 */

{
   FILE *fp, *fopen();

   quietflag = false;
   if (argc > 1 && strcmp("-q",argv[1])==0) {
        quietflag = true;
        argc--; argv++;
   }

   if (argc<2) {
	 if ((scanfile(stdin) == 0) && !quietflag)
	    VOID fprintf(stderr, "ident warning: no id keywords in input\n");
	exit(0);
   }

   while ( --argc > 0 ) {
      if ( (fp = fopen(*++argv, "r") ) == NULL ) {
         VOID fprintf(stderr,  "ident error: can't open %s\n", *argv);
         continue;
      } else {
         VOID printf( "%s:\n", *argv);   /*  print file name  */
	 if ((scanfile(fp) == 0) && !quietflag)
	    VOID fprintf(stderr, "ident warning: no id keywords in %s\n", *argv);
	 if (argc>1) putchar('\n');
	 VOID fclose(fp);
      }
   }
   exit(0);
}


int scanfile(file)
FILE * file;
/* Function: scan an open file with descriptor file for keywords.
 * Returns the number of matches.
 */
{
   register int matchcount;
   register int c;


   matchcount = 0;
   while( (c=getc(file)) != EOF) {
      if ( (char)c==KDELIM)
	 matchcount += match(file);
   }
   return matchcount;
}



match(fp)   /* group substring between two KDELIM's; then do pattern match */
FILE *fp;
{
   char line[keyvallength];
   register int c;
   register char * tp;

   tp = line;
   while( (c = getc(fp)) != KDELIM ) {
      *tp++ = c;
      if ( c==EOF || c=='\n' || tp>= line+keyvallength-2)
         return(0);
   }
   *tp++ = c;     /*append trailing KDELIM*/
   *tp   = '\0';
   if (trymatch(line,true)!=Nomatch) {
        VOID fprintf(stdout,"     $%s\n",line);
        return(1);
   } else {
      /* no match; put trailing KDELIM back into input */
      VOID ungetc(c,fp );
      return(0);
   }
}
